Clean up Stopwatch a bit

stephentoub opened the pull request:
- Remove unnecessary Reset call from ctor
- Remove unnecessary branch for _elapsed < 0 (Stop is now more inlineable)
- Remove some defunct comments
- Clean up style of Stop to match that of Start

Closes https://github.com/dotnet/runtime/issues/66734
Related to https://github.com/dotnet/runtime/issues/111829


| Method | Toolchain         | Mean     | Ratio | Code Size | Allocated | Alloc Ratio |
|------- |------------------ |---------:|------:|----------:|----------:|------------:|
| Time   | \main\corerun.exe | 38.51 ns |  1.00 |     363 B |      40 B |        1.00 |
| Time   | \pr\corerun.exe   | 30.35 ns |  0.79 |     156 B |         - |        0.00 |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Diagnostics;
using System.Runtime.CompilerServices;

BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

[DisassemblyDiagnoser]
[MemoryDiagnoser(false)]
public partial class Tests
{
    [Benchmark]
    public TimeSpan Time()
    {
        Stopwatch sw = Stopwatch.StartNew();
        Nop();
        sw.Stop();
        return sw.Elapsed;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void Nop() { }
}
```

stephentoub committed:
Clean up Stopwatch a bit

- Remove unnecessary Reset call from ctor
- Remove unnecessary branch for _elapsed < 0
- Remove some defunct comments
- Clean up style of Stop to match that of Start

paulbartrum commented:
Are the `Debug.Assert(IsHighResolution);` asserts still needed? These will always be true as far as I can tell.

stephentoub committed:
More stylistic cleanup

- Use expression-bodied properties
- Remove unnecessary Debug.Asserts
- Remove private method whose impl could just have been that of an existing public property
- Change a private GetXx method to an Xx property.
- Remove some type names that could be inferred

stephentoub commented:
> Are the `Debug.Assert(IsHighResolution);` asserts still needed? These will always be true as far as I can tell.

They're not needed.

MihaZupan reviewed (approved):
Nice

stephentoub commented:
/ba-g unrelated infra

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs:
```diff
@@ -53,29 +43,16 @@ public void Stop()
+                _elapsed += GetTimestamp() - _startTimeStamp;
                 _isRunning = false;
-
-                if (_elapsed < 0)
-                {
-                    // When measuring small time periods the Stopwatch.Elapsed*
-                    // properties can return negative values.  This is due to
-                    // bugs in the basic input/output system (BIOS) or the hardware
-                    // abstraction layer (HAL) on machines with variable-speed CPUs
-                    // (e.g. Intel SpeedStep).
-
-                    _elapsed = 0;
-                }
```
I believe we still support CPUs that could have this issue on Linux, although such processors are increasingly rare.

The Intel Architecture Manual covers that RDTSC is tied to the processor frequency for Pentium M processors (family [06H], models [09H, 0DH]); for Pentium 4 processors, Intel Xeon processors
(family [0FH], models [00H, 01H, or 02H]); and for P6 family processors

For a constant rate that is not tied to the CPU frequency, you instead need: Pentium 4 processors, Intel Xeon processors (family [0FH], models [03H and higher]); for Intel Core Solo
and Intel Core Duo processors (family [06H], model [0EH]); for the Intel Xeon processor 5100 series and Intel
Core 2 Duo processors (family [06H], model [0FH]); for Intel Core 2 and Intel Xeon processors (family [06H],
DisplayModel [17H]); for Intel Atom processors (family [06H], DisplayModel [1CH]).

But even more so, it is technically dependent on querying that the CPU supports `Invariant TSC` by querying `CPUID 0x8000_0007` and then reading `EDX bit 8`.

`QPC` on Windows and `CLOCK_MONOTONIC` on Unix is supposed to normalize this, but historically there have been bugs.

tannergooding reviewed (commented):
LGTM.

TonyValenti commented:
Regarding negative values, I *THINK* I have observed this in the past and attributed it to either a daylight savings time adjustment or windows time synchronization. Maybe those are no longer issues but I wanted to share.

stephentoub commented:
It's worth reiterating that this special-casing was _only_ in Stop. It didn't apply if you did:
```C#
sw.Start();
...;
TimeSpan elapsed = sw.Elapsed;
```
only if you did:
```C#
sw.Start();
...;
sw.Stop();
TimeSpan elapsed = sw.Elapsed;
```
That's true on .NET Framework as well.

Given that, even if there are still some machines in circulation where this _could_ manifest, I'm comfortable removing this one place it was being special-cased.
